### ISSUE TITLE
Bug: renderToString & Inconsistent Behavior Updating State

### ISSUE THREAD

**Issue by michaelyuen** [facebook/react#18294]
React version: 16.12.10

## Overview
I do not know if this is a bug or expected behavior; I'm hoping to receive that answer. I've searched docs, issues, etc. and haven't found a clear answer on the behavior expected by `renderToString` in this scenario.

## Steps To Reproduce
This is not a fully implemented SSR example. The main thing to focus on is the result of `renderToString`. The pre-existing `render` remains for comparison purposes.

1. Open codesandbox
2. Review console logs

Link to code example:
1. [React SSR + Context](https://codesandbox.io/s/react-ssr-context-z4tqi)
2. [React SSR + Passed State](https://codesandbox.io/s/react-ssr-passed-state-yocro)

## The current behavior
**TLDR:** renderToString will re-render a component due to a state change, but only if the state is defined in that component.

In the first example (React SSR + Context), review the `App.js` file and notice there are two sets of state (value and setter), one defined _in_ App (this is important) and one defined in `context.js` and consumed with a custom hook.

The code is written to update the state immediately in order to demonstrate what _I think_ is an inconsistency. Reviewing the logged `htmlString`, notice that it logs the "this will be server-rendered" html, but not the "this will NOT be server-rendered". To me, this means App's state change has been processed and re-rendered, but the state in context has not changed or triggered the consumer (App) to re-render. My big question here is **why?**

In the second example (React SSR + Passed State), the setup is quite similar, but takes an even more direct approach by sharing state through props. This was intended to validate the theory that the previously noted observation is related to Context, however, this example demonstrates the same behavior.

## The expected behavior
**TLDR:** renderToString will re-render a component due to a state change defined in a component and updated in another component.

Seeing that the html string produced by `renderToString` indicates that the state defined and updated in `App` was processed, I expected the same behavior for the state defined in `TestProvider` and updated in `App`.

**Comment by mertxe78** [comment]
_ ** # ## **** ** _

**Comment by gaearon** [comment]
>TLDR: renderToString will re-render a component due to a state change, but only if the state is defined in that component.

This is the expected behavior.

>TLDR: renderToString will re-render a component due to a state change defined in a component and updated in another component.

This can't possibly work unless we buffer all HTML, negating the possible benefits of streaming. Cause you won't know that something later might cause a state to update.

Even if we did support that, it would be slow because we would have to add a lot of machinery to track state updates during a server render. So we won't be doing this.

We have a warning since React 16.13 for this. If you update your sandbox, you will see it.
